rfcs/20220329-Extending-CUDA-backend-of-oneDNN-Primitives/README.md

```diff
@@ -0,0 +1,190 @@
+    - `CUDNN_REDUCE_TENSOR_MIN` 
+    - `CUDNN_REDUCE_TENSOR_MUL`
+    - `CUDNN_REDUCE_TENSOR_ADD` 
+    - `CUDNN_REDUCE_TENSOR_AVG` 
+
+- Supported Data Types:  f32, s8 
+
+ 
+
+#### Implementation Limitations 
+
+- Lack of support for the algorithms of  
+
+    - `reduction_norm_lp_max` 
+
+    - `reduction_norm_lp_sum` 
+
+    - `reduction_norm_lp_power_p_max` 
+
+    - `reduction_norm_lp_power_p_sum` 
+
+- No support for bf16 datatype. 
+
+ 
+
+#### Proposed Solutions 
+
+- To achieve the functionality of `reduction_norm_lp_power_p_max` and `reduction_norm_lp_power_p_sum` algorithms in cuDNN, `CUDNN_REDUCE_TENSOR_NORM1` algorithm mode of `cudnnReduceTensor` can be used in combination with `CUDNN_REDUCE_TENSOR_MAX` and `CUDNN_REDUCE_TENSOR_ADD` respectively. 
+
+    - However, this workaround doesn't provide the flexibility to use arbitrary value for power `p` other than 1.  
+
+- Similarly, the functionality of `reduction_norm_lp_max` and `reduction_norm_lp_sum` algortihms can be implemented using `CUDNN_REDUCE_TENSOR_NORM1` algorithm mode of `cudnnReduceTensor` in combination with `CUDNN_REDUCE_TENSOR_MAX` and `CUDNN_REDUCE_TENSOR_ADD` respectively. 
+
+    - However, this workaround doesn't provide the flexibility to use arbitrary value for power `p` other than 2.  
+
+- As an alternative to the above workarounds, the functionality of the above algorithms can be implemented through a custom kernel in SYCL. 
+
+ 
+
+### Recurrent Neural Networks 
+
+- RNN primitive in CUDA backend can be implemented using `cudnnRNNForward` and `cudnnRNNBackward` primitives of cuDNN. 
+
+- cuDNN's primitives only support two algorithm of RNN. 
+
+    - `CUDNN_LSTM` 
+
+    - `CUDNN_GRU` 
+
+- Supported Data Types: f16, f32 
+
+ 
+
+#### Propagation 
+
+Supported propagation kinds: Forward Training/Inference and Backward 
+
+ 
+
+#### Implementation Limitations 
+
+- No support for following algorithms in cuDNN. 
+
+    - `augru` 
+
+    - `lbr_augru` 
+
+    - `lbr_gru` 
+
+    - `vanilla_rnn` 
+
+- No support for U8, bf16 datatypes. 
+
+ 
+
+#### Proposed Solution 
+
+The unsupported algorithms of `augru`, `lbr_augru`, `lbr_gru`, `vanilla_rnn` in cuDNN can be implemented using custom kernels, per the mathematical expressions of the algorithms. 
+
+ 
+
+### PRelu  
+
+There is no equivalent primitive API in cuDNN for PRelu. 
+
+ 
+
+#### Propagation 
+
+Forward Training/Inference and Backward 
+
+ 
+
+#### Implementation Limitations 
+
+Lack of equivalent support for PRelu primitive in cuDNN. 
+
+ 
+
+#### Proposed Solution 
+
+- PRelu primitive can be implemented using cuDNN with `cudnnActivationForward()` and `cudnnActivationBackward()`. 
+
+    - Using `CUDNN_ACTIVATION_RELU` algorithm mode, it is possible to implement LEAKY RELU with a constant scalar alpha value, but not PRelu with a tensor alpha value. 
+
+    - However, the functionality of PRelu can be accomplished using `CUDNN_ACTIVATION_RELU` algorithm mode in combination with `cudnnOpTensor()`.
```

Comment:
Can you please briefly explain the idea of combining classic ReLU with `cudnnOpTensor`? Also, do you know how to support it for backward pass where alpha (weights) is a learnable parameter?

Also, the primitive supports broadcasting for weights that should also be considered.

Reply:
The implementation proposed with cudnnOpTensor and RELU is very challenging as it involves multiple primitives i.e., multiple instances of cudnnOpTensor. The backward pass can be implemented using the tensor operations - MIN & MUL. Since this approach looks less efficient, we are planning for custom kernel implementation.